Close socket on panic()

**Is your feature request related to a problem? Please describe.**
When bnsd failed to start due to incorrect genesis file the unix socket file was still existing and required further clean to restart the app.

**Describe the solution you'd like**
Proper house-keeping would close the socket.

**Describe alternatives you've considered**
Document this behaviour

**Additional context**
Example of a panic stacktrace with v0.18.0:
```
panic: invalid escrow at position: 0 : destination: value is empty

goroutine 33 [running]:
github.com/iov-one/weave/app.(*StoreApp).InitChain(0xc0004d6b40, 0xbc87a66, 0xed4ae801f, 0x0, 0xc0000d54e0, 0x11, 0xc000086300, 0xc0004d8150, 0x1, 0x1, ...)
	/home/travis/gopath/src/github.com/iov-one/weave/app/store.go:317 +0x16d
github.com/tendermint/tendermint/abci/server.(*SocketServer).handleRequest(0xc000196be0, 0xc0004d1bf0, 0xc00006df20)
	/home/travis/gopath/pkg/mod/github.com/tendermint/tendermint@v0.31.5/abci/server/socket_server.go:193 +0x81d
github.com/tendermint/tendermint/abci/server.(*SocketServer).handleRequests(0xc000196be0, 0xc00006de00, 0xf8d8a0, 0xc000096040, 0xc00006df20)
	/home/travis/gopath/pkg/mod/github.com/tendermint/tendermint@v0.31.5/abci/server/socket_server.go:163 +0xe4
created by github.com/tendermint/tendermint/abci/server.(*SocketServer).acceptConnectionsRoutine
	/home/travis/gopath/pkg/mod/github.com/tendermint/tendermint@v0.31.5/abci/server/socket_server.go:119 +0x26a
```

--- a/commands/server/start.go
+++ b/commands/server/start.go
@@ -72,17 +72,28 @@ func StartCmd(gen AppGenerator, logger log.Logger, home string, args []string) e
 	if err != nil {
 		return errors.Wrap(err, "failed to create a listener")
 	}
+
 	svr.SetLogger(logger.With("module", "abci-server"))
+
+	done := make(chan bool)
+	cleanupCallback := func() {
+		// Cleanup
+		_ = svr.Stop()
+		done <- true
+	}
+
+	cmn.TrapSignal(logger, cleanupCallback)
+
+	defer func() {
+		if r := recover(); r != nil {
+			cleanupCallback()
+		}
+	}()
+
 	err = svr.Start()
 	if err != nil {
 		return errors.Wrap(err, "failed to start a server")
 	}
-	done := make(chan bool)
-	cmn.TrapSignal(logger, func() {
-		// Cleanup
-		_ = svr.Stop()
-		done <- true
-	})
 
 	// wait forever
 	<-done

--- a/commands/server/start.go
+++ b/commands/server/start.go
@@ -85,7 +85,8 @@ func StartCmd(gen AppGenerator, logger log.Logger, home string, args []string) e
 	cmn.TrapSignal(logger, cleanupCallback)
 
 	defer func() {
-		if r := recover(); r != nil {
+		if err := recover(); err != nil {
+			logger.Error(err)
 			cleanupCallback()
 		}
 	}()

--- a/commands/server/start.go
+++ b/commands/server/start.go
@@ -2,7 +2,6 @@ package server
 
 import (
 	"flag"
-
 	"github.com/iov-one/weave/coin"
 	"github.com/iov-one/weave/errors"
 	"github.com/tendermint/tendermint/abci/server"
@@ -86,7 +85,7 @@ func StartCmd(gen AppGenerator, logger log.Logger, home string, args []string) e
 
 	defer func() {
 		if err := recover(); err != nil {
-			logger.Error(err)
+			logger.Error("recovered from panic", "err", err)
 			cleanupCallback()
 		}
 	}()

